### ISSUE TITLE
Seurat Object command slot enormous, possibly holding serialized seurat objects?

### ISSUE THREAD

**Issue by bbimber** [satijalab/seurat#5343]
Hello,

We are working with large seurat objects, getting into several hundred thousand cells. As a matter of practicality, we are subsetting these object based on characteristics, like cluster or cell type, and then operating on the subset, smaller objects. After subset, in many cases the object size is not reducing to a degree I'd expect, I'm not not yet sure why. Below is how we're inspecting an object, and I am wondering if you have comments or suggestions on ways we could debug this. 

The punch line is that the commands slot is enormous and might be holding serialized seurat objects. FWIW, we often call seurat using do.call(). That's the only atypical thing we might be doing. 

Here's the full summary of what we did: Taking an example with an input object with two assays and ~5000 cells, we ran this:

```
InspectSeurat <- function(seuratObj) {
  print(paste0('Seurat object size: ', format(object.size(seuratObj), units = 'auto')))
  nCells <- ncol(seuratObj)
  if (nrow(seuratObj@meta.data) != nCells) {
    print(paste0('Metadata rows not equal to ncol: ', nCells, ' / ', nrow(seuratObj@meta.data)))
  }
    
  for (assayName in names(seuratObj@assays)) {
    print(paste0('Assay: ', assayName))
    for (slotName in c('counts', 'data', 'scale.data')) {
      dat <- Seurat::GetAssayData(seuratObj, assay = assayName, slot = slotName)
      print(paste0('slot: ', slotName, ', size: ', format(object.size(x = dat), units = 'auto')))
      
      if (!is(dat, 'sparseMatrix')) {
        print(paste0('Non-sparse! Assay: ', assayName, ', slot: ', slotName))
      }
      
      if (ncol(dat) > 0 && ncol(dat) != nCells) {
        print(paste0('Assay cols not equal to ncol(object): ', nCells, ' / ', ncol(dat)))
      }
    }  
  }
  
  print('Reductions:')
  for (reductionName in names(seuratObj@reductions)) {
    print(paste0(reductionName, ', size: ', format(object.size(x = seuratObj@reductions[reductionName]), units = 'auto')))
  }
  
  print('All slots:')
  for (slotName in slotNames(seuratObj)) {
    val <- object.size(slot(seuratObj, slotName))
    if (val > 1000) { 
      print(paste0(slotName, ': ', format(val, units = 'auto')))
    }
  }
  
  if (nrow(seuratObj@meta.data) != nCells) {
    print(paste0('Metadata rows not equal to ncol: ', nCells, ' / ', nrow(seuratObj@meta.data)))
  }
}

InspectSeurat(seuratObj)

```

The result:

```
[1] "Seurat object size: 2.1 Gb"
[1] "Assay: RNA"
[1] "slot: counts, size: 88.4 Mb"
[1] "slot: data, size: 88.4 Mb"
[1] "slot: scale.data, size: 87 Mb"
[1] "Non-sparse! Assay: RNA, slot: scale.data"
[1] "Assay: ADT"
[1] "slot: counts, size: 2.3 Mb"
[1] "slot: data, size: 2.4 Mb"
[1] "slot: scale.data, size: 216 bytes"
[1] "Non-sparse! Assay: ADT, slot: scale.data"
[1] "Reductions:"
[1] "pca, size: 5.1 Mb"
[1] "tsne, size: 563.7 Kb"
[1] "umap, size: 563.8 Kb"
[1] "All slots:"
[1] "assays: 274.3 Mb"
[1] "meta.data: 1.3 Mb"
[1] "active.ident: 496.2 Kb"
[1] "graphs: 7.5 Mb"
[1] "reductions: 6.2 Mb"
[1] "commands: 1.8 Gb"
```

Commands is 1.8GB? It appears to be holding serialized seurat objects? 

and then we subset to just 100 cells

```
seuratObj3 <- subset(seuratObj, cells = colnames(seuratObj)[1:100])
print(ncol(seuratObj3))
InspectSeurat(seuratObj3)
```

The assays and other slots reduce somewhat, but commands is still enormous (since subset likely doesnt reduce this).

**Comment by bbimber** [comment]
After more review, the problem in our specific case is RunUMAP.RNA.pca, and call.string, which is 1.5GB. This is created using do.call(), and probably holds the seurat object.

**Comment by timoast** [comment]
This is related to the use of `do.call()`, please see https://github.com/satijalab/seurat/issues/3312

**Comment by bbimber** [comment]
OK. have you considered exec() vs. invoke()? The latter is soft-deprecated in favor of exec().

Second, is there a public way to clear seuratObj@commands, or is:

seuratObj@commands <- list()

the right way

**Comment by timoast** [comment]
I have not tried `rlang::exec()`, if you try it please let me know if it works.

There's no "official" way to remove the command information, but you could assign it to a new list as you showed above.

**Comment by bbimber** [comment]
tentatively, exec() seems to work. I went from:

seuratObj <- do.call(RunUMAP, umapArgs)

to:

seuratObj <- rlang:::exec(RunUMAP, !!!umapArgs)

**Comment by bbimber** [comment]
Note: I just see that it doesnt error. I didnt inspect object size yet

**Comment by bbimber** [comment]
For future people reading this, 

seuratObj@commands <- list()

has the one problem that certain seurat functions expect specific commands to have been run, and therefore simply discarding the command line by itself is not a solution unless you're re-running everything.

**Comment by bbimber** [comment]
OK, so my observation is that rlang::exec() makes no difference on logging large strings to @commands relative to do.call(). We're going to implement this after any instance where we use exec():

```
.ClearSeuratCommands <- function(seuratObj, maxSize = 500000) {
	for (commandName in names(seuratObj@commands)) {
		val <- object.size(x = slot(seuratObj@commands[[commandName]], 'call.string'))
		if (val > maxSize) {
			print(paste0('Clearing call.string for: ', commandName, '. size: ', format(val, units = 'auto')))
			slot(seuratObj@commands[[commandName]], 'call.string') <- ''
		}
	}

	return(seuratObj)
} 
```

which will at least mitigate the problem.

**Comment by bbimber** [comment]
@timoast From what I can see, invoke (which is soft-deprecated) gives a good result; however, rlang::exec (which is the replacement for invoke), and do.call result in the whole object being logged. Below is a small repro case illustrating this. You can close this thread if you're not going to follow up, but I thought I'd pass this along:

```


dat <- matrix(sample(1:100, 1000, replace = T), ncol = 100)
rownames(dat) <- LETTERS[1:nrow(dat)]
colnames(dat) <- paste0('Cell-', 1:ncol(dat))
seuratObj <- CreateSeuratObject(dat, project = 'Project1')

reportCommands <- function(seuratObj) {
  for (cn in names(seuratObj@commands)) {
    val <- seuratObj@commands[[cn]]@call.string
    print(paste0(cn, ': ', length(val)))
    if (length(val) == 1) {
      print(val)
    } 
  }
}

doWork <- function(seuratObj, mode) {
  if (mode == 'do.call') {
    seuratObj <- do.call(NormalizeData, list(object = seuratObj, verbose = F))
  } else if (mode == 'basic') {
    seuratObj <- NormalizeData(seuratObj, verbose = F)
  } else if (mode == 'exec') {
    seuratObj <- rlang::exec(NormalizeData, !!!list(object = seuratObj, verbose = F))
  } else if (mode == 'invoke') {
    seuratObj <- rlang::invoke(NormalizeData, list(object = seuratObj, verbose = F))
  }
  
  return(seuratObj)
}

# Reasonable results
seuratObj <- doWork(seuratObj, mode = 'basic')
reportCommands(seuratObj)

seuratObj <- doWork(seuratObj, mode = 'invoke')
reportCommands(seuratObj)

# Entire object cached
seuratObj <- doWork(seuratObj, mode = 'do.call')
reportCommands(seuratObj)

seuratObj <- doWork(seuratObj, mode = 'exec')
reportCommands(seuratObj)

```

**Comment by timoast** [comment]
Thanks @bbimber for looking into this. I will close this for now and recommend using `rlang::invoke()` rather than `do.call()` as a workaround

**Comment by wtorous** [comment]
I recently encountered this same issue. Of the three possible options, `rlang::exec()` makes a smaller Seurat object than `do.call()` or `rlang::invoke()`. However, I still needed @bbimber's `ClearSeuratCommands` function for the pipeline the run reliably. @timoast Do you know of any fixes in the works (I am using 4.1.1)?

In the following image, I print the memory used to store a Seurat object through some standard steps. The first run is with direct calls to the Seurat functions, the second run uses `rlang::exec()`, and the third uses `do.call()` which cannot handle scaling. Note that the object after scaling is almost 900 Mb with `rlang::exec()` and less than 300 Mb with direct calls.

<img width="598" alt="full_memory_comparison" src="https://user-images.githubusercontent.com/38775272/202018234-d3ed731d-9c69-4d35-b42e-216e9313ec6a.png">
